cordova-plugin-outline/apple/vpn/PacketTunnelProvider.m

```diff
@@ -121,46 +121,38 @@ - (void)startTunnelWithOptions:(NSDictionary *)options
   // still be unusable, but at least the user will have a visual indication that Outline is the
   // culprit and can explicitly disconnect.
   long errorCode = noError;
-  __block NSError *err = nil;
   if (!isOnDemand) {
     ShadowsocksCheckConnectivity(self.hostNetworkAddress, [self.connection.port intValue],
-                                 self.connection.password, self.connection.method, &errorCode,
-                                 &err);
+                                 self.connection.password, self.connection.method, &errorCode, nil);
   }
-  __block ErrorCode clientErrorCode =
-      (errorCode == noError || errorCode == udpRelayNotEnabled) ? noError : errorCode;
-  if (clientErrorCode == noError) {
-    [self connectTunnel:[self getTunnelNetworkSettings]
-             completion:^(NSError *error) {
-               if (!error) {
-                 BOOL isUdpSupported =
-                     isOnDemand ? self.connectionStore.isUdpSupported : errorCode == noError;
-                 if (![self startTun2Socks:isUdpSupported]) {
-                   clientErrorCode = vpnStartFailure;
-                   err = [NSError errorWithDomain:NEVPNErrorDomain
-                                             code:NEVPNErrorConnectionFailed
-                                         userInfo:nil];
-                 } else {
-                   [self listenForNetworkChanges];
-                   [self.connectionStore save:connection];
-                   self.connectionStore.isUdpSupported = isUdpSupported;
-                   self.connectionStore.status = ConnectionStatusConnected;
-                 }
-               } else {
-                 clientErrorCode = vpnPermissionNotGranted;
-                 err = [NSError errorWithDomain:NEVPNErrorDomain
-                                           code:NEVPNErrorConnectionFailed
-                                       userInfo:nil];
-               }
-               [self execAppCallbackForAction:kActionStart errorCode:clientErrorCode];
-               completionHandler(err);
-             }];
-  } else {
-    [self execAppCallbackForAction:kActionStart errorCode:clientErrorCode];
-    completionHandler([NSError errorWithDomain:NEVPNErrorDomain
-                                          code:NEVPNErrorConnectionFailed
-                                      userInfo:nil]);
+  if (errorCode != noError && errorCode != udpRelayNotEnabled) {
+    [self execAppCallbackForAction:kActionStart errorCode:errorCode];
+    return completionHandler([NSError errorWithDomain:NEVPNErrorDomain
+                                                 code:NEVPNErrorConnectionFailed
+                                             userInfo:nil]);
   }
+
+  [self connectTunnel:[self getTunnelNetworkSettings]
+           completion:^(NSError *error) {
```

Comment:
Nit: It looks like you added `_Nullable` to the other `completion`s, so this one should probably be parallel.

Reply:
Done, good catch.